heft-plugins/heft-webpack4-plugin/src/Webpack4Plugin.ts

```diff
@@ -0,0 +1,283 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { AsyncParallelHook, AsyncSeriesWaterfallHook } from 'tapable';
+import type * as TWebpack from 'webpack';
+import type TWebpackDevServer from 'webpack-dev-server';
+import { LegacyAdapters } from '@rushstack/node-core-library';
+import type {
+  HeftConfiguration,
+  IHeftTaskSession,
+  IHeftTaskCleanHookOptions,
+  IHeftTaskPlugin,
+  IHeftTaskRunHookOptions,
+  IScopedLogger
+} from '@rushstack/heft';
+
+import type {
+  IWebpackConfiguration,
+  IWebpackConfigurationWithDevServer,
+  IWebpack4PluginAccessor,
+} from './shared';
+import { WebpackConfigurationLoader } from './WebpackConfigurationLoader';
+
+export interface IWebpack4PluginOptions {
+  devConfigurationPath: string | undefined;
+  configurationPath: string | undefined;
+}
+
+/**
+ * @public
+ */
+export const PLUGIN_NAME: 'Webpack4Plugin' = 'Webpack4Plugin';
+const WEBPACK_DEV_SERVER_PACKAGE_NAME: 'webpack-dev-server' = 'webpack-dev-server';
+const WEBPACK_DEV_SERVER_ENV_VAR_NAME: 'WEBPACK_DEV_SERVER' = 'WEBPACK_DEV_SERVER';
+
+/**
+ * @internal
+ */
+export default class Webpack4Plugin implements IHeftTaskPlugin<IWebpack4PluginOptions> {
+  private _webpack: typeof TWebpack | undefined;
+  private _webpackConfiguration: IWebpackConfiguration | undefined;
+
+  public readonly accessor: IWebpack4PluginAccessor = {
+    onConfigureWebpackHook: new AsyncSeriesWaterfallHook(['webpackConfiguration']),
+    onAfterConfigureWebpackHook: new AsyncParallelHook(['webpackConfiguration']),
+    onEmitStatsHook: new AsyncParallelHook(['webpackStats'])
+  };
+
+  public apply(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    options: IWebpack4PluginOptions
+  ): void {
+    // These get set in the run hook and used in the onConfigureWebpackHook
+    let production: boolean;
+    let serveMode: boolean;
+    let watchMode: boolean;
+
+    // Provide the initial webpack configuration by tapping the hook
+    this.accessor.onConfigureWebpackHook.tapPromise(
+      PLUGIN_NAME,
```

Comment:
This tap will run first (since it's the first thing registered) unless someone explicitly tells their tap to run earlier. It'd make sense to specifically tell tapable to run this last.

Reply:
Applies to the Webpack 5 version too.

Reply:
Running it last would ensure that no one could modify the configuration, which I thought was part of the reason this hook existed? I kept it running without an explicit stage number because that allows other plugins to run either before or after, depending on what they set `stage` to.

Reply:
Spoke offline, agreed we should break this into multiple different hooks to clarify use cases.